File: card-common/src/main/java/com/jean202/cardmizer/common/Money.java
package com.jean202.cardmizer.common;

import java.util.Objects;

public record Money(long amount) implements Comparable<Money> {
    public static final Money ZERO = new Money(0);

    public Money {
        if (amount < 0) {
            throw new IllegalArgumentException("Money amount must be zero or positive");
        }
    }

    public static Money won(long amount) {
        return new Money(amount);
    }

    public Money add(Money other) {
        Objects.requireNonNull(other, "other must not be null");
        return new Money(Math.addExact(amount, other.amount));
    }

    public Money subtract(Money other) {
        Objects.requireNonNull(other, "other must not be null");
        long result = Math.subtractExact(amount, other.amount);
        if (result < 0) {
            throw new IllegalArgumentException("Money amount must not become negative");
        }
        return new Money(result);
    }

    public boolean isGreaterThanOrEqual(Money other) {
        Objects.requireNonNull(other, "other must not be null");
        return amount >= other.amount;
    }

    @Override
    public int compareTo(Money other) {
        Objects.requireNonNull(other, "other must not be null");
        return Long.compare(amount, other.amount);
    }
}

File: card-common/src/main/kotlin/com/jean202/cardmizer/common/Money.kt
package com.jean202.cardmizer.common

@JvmInline
value class Money private constructor(val amount: Long) : Comparable<Money> {

    companion object {
        val ZERO = Money(0L)

        @JvmStatic
        fun won(amount: Long): Money {
            require(amount >= 0L) { "Money amount must be zero or positive" }
            return Money(amount)
        }
    }

    operator fun plus(other: Money): Money = Money(Math.addExact(amount, other.amount))

    operator fun minus(other: Money): Money {
        val result = Math.subtractExact(amount, other.amount)
        require(result >= 0L) { "Money amount must not become negative" }
        return Money(result)
    }

    fun add(other: Money): Money = plus(other)
    fun subtract(other: Money): Money = minus(other)

    fun isGreaterThanOrEqual(other: Money): Boolean = amount >= other.amount

    override fun compareTo(other: Money): Int = amount.compareTo(other.amount)
}
